doc/sphinx/practical-tools/utilities.rst

```diff
@@ -559,43 +629,73 @@ To build, say, two targets foo.vo and bar.vo in parallel one can use
 
 .. note::
 
-  ``make foo.vo bar.vo -j`` has a different meaning for the make
+  ``make foo.vo bar.vo -j`` has a different meaning for the ``make``
   utility, in particular it may build a shared prerequisite twice.
 
-
-.. note::
-
-  For users of coq_makefile with version < 8.7
-
-  + Support for "subdirectory" is deprecated. To perform actions before
-    or after the build (like invoking ``make`` on a subdirectory) one can hook
-    in pre-all and post-all extension points.
-  + ``-extra-phony`` and ``-extra`` are deprecated. To provide additional target
-    (``.PHONY`` or not) please use ``CoqMakefile.local``.
-
 .. note::
 
   Due to limitations with the compilation chain, makefiles generated
   by ``coq_makefile`` won't correctly compile OCaml plugins with OCaml
   < 4.07.0 when using more than one job (``-j N`` for ``N > 1``).
 
+
+The grammar of _CoqProject
+++++++++++++++++++++++++++
+The parser of ``coq_makefile`` uses the following syntax:
+
+  .. prodn::
+     CoqProject ::= {* {| @blank | @comment | @quoted_string | @unquoted_string } }
+     blank ::= {| space | horizontal_tab | newline }
+     comment ::= # {* comment_char } newline
+     quoted_string ::= " {* quoted_char } "
+     unquoted_string ::= string_start_char {* unquoted_char }
+
+Where the following definitions apply:
+
+* :n:`space`, :n:`horizontal_tab` and :n:`newline` stand for the corresponding
+  ASCII characters.
+* :n:`comment_char` is the set of all characters except :n:`newline`.
+* :n:`quoted_char` is the set of all characters except ``"``.
+* :n:`string_start_char` is the set of all characters except those that match :n:`@blank`, or are ``"`` or ``#``.
+* :n:`unquoted_char` is the set of all characters except those that match :n:`@blank` or are ``#``.
+
+The parser produces a list of strings in the same order as they were
+encountered in ``_CoqProject``. All :n:`@blank` and :n:`@comment` tokens
+get ignored and the double quotes of :n:`@quoted_string` tokens are removed as
+well.
+
+The current approach has a few limitations: Double quotes are only special
+characters at the start of a string. And for lack of an escaping mechanism, it
+is currently impossible to encode the following kinds of strings in a
+``_CoqProject`` file:
+
+* strings starting with ``"``
+* strings starting with ``#`` and containing ``"``
+* strings containing both whitespace and ``"``
+
+The list of strings is then processed as follows: If the first string of the
+list is a known option, interpret the next strings as arguments of this option
+and remove these strings from the list. If the first string of the list is not
```

Comment:
"next strings" - on the same line?

Unclear: "removing these strings" meaning it won't be passed to coqc?  The text should be clear about what is done with the resulting list of strings (i.e. it's passed to coqc IIUC).

Reply:
I tried to describe `process_cmd_line` using imperative wording. I'll try to think of better wording.

Reply:
It's OK to use the passive voice at times.  The imperative form works when the subject is the reader, e.g. "[You] eat your broccoli"

Reply:
I wasn't clear. I meant "imperative vs. functional programming", not in the grammatical sense.